altair/utils/schemapi.py

```diff
@@ -113,22 +134,98 @@ def _get_errors_from_spec(
+        validator_kwargs["resolver"] = (
+            jsonschema.RefResolver.from_schema(rootschema)
+            if rootschema is not None
+            else None
+        )
+
     validator = validator_cls(schema, **validator_kwargs)
     errors = list(validator.iter_errors(spec))
     return errors
 
 
+def _get_json_schema_draft_url(schema: dict) -> str:
+    return schema.get("$schema", _DEFAULT_JSON_SCHEMA_DRAFT_URL)
+
+
+def _use_referencing_library() -> bool:
+    """In version 4.18.0, the jsonschema package deprecated RefResolver in
+    favor of the referencing library."""
+    jsonschema_version_str = importlib_version("jsonschema")
+    return Version(jsonschema_version_str) >= Version("4.18")
+
+
+def _prepare_references_in_schema(schema: Dict[str, Any]) -> Dict[str, Any]:
+    # Create a copy so that $ref is not modified in the original schema in case
+    # that it would still reference a dictionary which might be attached to
+    # an Altair class _schema attribute
+    schema = copy.deepcopy(schema)
+
+    def _prepare_refs(d: Dict[str, Any]) -> Dict[str, Any]:
+        """Add _VEGA_LITE_ROOT_URI in front of all $ref values. This function
+        recursively iterates through the whole dictionary."""
+        for key, value in d.items():
+            if key == "$ref":
+                d[key] = _VEGA_LITE_ROOT_URI + d[key]
+            else:
+                # $ref values can only be nested in dictionaries or lists
+                # as the passed in `d` dictionary comes from the Vega-Lite json schema
+                # and in json we only have arrays (-> lists in Python) and objects
+                # (-> dictionaries in Python) which we need to iterate through.
+                if isinstance(value, dict):
+                    d[key] = _prepare_refs(value)
+                elif isinstance(value, list):
+                    prepared_values = []
+                    for v in value:
+                        if isinstance(v, dict):
+                            v = _prepare_refs(v)
+                        prepared_values.append(v)
+                    d[key] = prepared_values
+        return d
+
+    schema = _prepare_refs(schema)
+    return schema
+
+
+# We do not annotate the return value here as the referencing library is not always
+# available and this function is only executed in those cases.
+def _get_referencing_registry(
+    rootschema: Dict[str, Any], json_schema_draft_url: Optional[str] = None
+):
+    # Referencing is a dependency of newer jsonschema versions, starting with the
+    # version that is specified in _use_referencing_library and we therefore
+    # can expect that it is installed if the function returns True.
+    import referencing
+    import referencing.jsonschema
+
+    if json_schema_draft_url is None:
+        json_schema_draft_url = _get_json_schema_draft_url(rootschema)
+
+    specification = referencing.jsonschema.specification_with(json_schema_draft_url)
```

Comment:
Thanks! Yeah that's what I originally did but not all schemas have a schema url attached so I had to revert to this. However, all schemas passed here are parts of the full Vega-Lite schema and use the same specification which is the one retrieved by `_get_json_schema_draft_url`.

Reply:
How often is there no `$schema` defined when there should be? Is this something to be raised in the VL repo so we can include a reference as comment to simplify this later?

Reply:
I think the VL schema is correct. `$schema` is defined once on the top-level which is enough as it is one JSON schema. In the code generation part in Altair, the generated Python classes have the `_schema` attribute set which refers to the corresponding definition in the VL schema. This attribute is set [here](https://github.com/altair-viz/altair/blob/main/tools/generate_schema_wrapper.py#L367) and you can see an example [here](https://github.com/altair-viz/altair/blob/main/altair/vegalite/v5/schema/core.py#L568). We don't set `$schema` in these generated dictionaries (`_schema`) which I think is fine as it would be the same across all of them. `_schema` is used in a few places such as during validation and for the property setter functionality, see [with_property_setters](https://github.com/altair-viz/altair/blob/main/altair/utils/schemapi.py#L1220).

Reply:
Thanks for the clarification @binste! I remember you mentioned something about this in relation to the property setter before, but I could not find this comment anymore.